ledger/src/blockstore_processor.rs

```diff
@@ -262,40 +262,111 @@ fn execute_batches_internal(
+                        }
+                        result
+                    },
+                    (),
+                    "execute_batch",
+                );
+
+                let thread_index = PAR_THREAD_POOL.current_thread_index().unwrap();
+                execution_timings_per_thread
+                    .write()
+                    .unwrap()
+                    .entry(thread_index)
+                    .and_modify(|thread_execution_time| {
+                        let ThreadExecuteTimings {
+                            total_thread_us,
+                            total_transactions_executed,
+                            execute_timings: total_thread_execute_timings,
+                        } = thread_execution_time;
+                        *total_thread_us += execute_batches_time.as_us();
+                        *total_transactions_executed += transaction_count;
+                        total_thread_execute_timings
+                            .saturating_add_in_place(ExecuteTimingType::TotalBatchesLen, 1);
+                        total_thread_execute_timings.accumulate(&timings);
+                    })
+                    .or_insert(ThreadExecuteTimings {
+                        total_thread_us: execute_batches_time.as_us(),
+                        total_transactions_executed: transaction_count,
+                        execute_timings: timings,
+                    });
+                result
+            })
+            .collect()
+    });
+    execute_batches_elapsed.stop();
+
+    cumulative_execute_timings
+        .saturating_add_in_place(ExecuteTimingType::TotalBatchesLen, batches.len() as u64);
+    cumulative_execute_timings.saturating_add_in_place(ExecuteTimingType::NumExecuteBatches, 1);
+    saturating_add_assign!(
+        confirmation_timing.execute_batches_us,
+        execute_batches_elapsed.as_us()
+    );
+
+    let mut current_max_thread_execution_time: Option<ThreadExecuteTimings> = None;
+    for (_, thread_execution_time) in execution_timings_per_thread
+        .into_inner()
+        .unwrap()
+        .into_iter()
+    {
+        let ThreadExecuteTimings {
+            total_thread_us,
+            execute_timings,
+            ..
+        } = &thread_execution_time;
+        cumulative_execute_timings.accumulate(execute_timings);
+        if *total_thread_us
+            > current_max_thread_execution_time
+                .as_ref()
+                .map(|thread_execution_time| thread_execution_time.total_thread_us)
+                .unwrap_or(0)
+        {
+            current_max_thread_execution_time = Some(thread_execution_time);
+        }
     }
 
+    if let Some(current_max_thread_execution_time) = current_max_thread_execution_time {
+        end_to_end_execute_timings.accumulate(&current_max_thread_execution_time);
+        end_to_end_execute_timings
+            .execute_timings
+            .saturating_add_in_place(ExecuteTimingType::NumExecuteBatches, 1);
+    };
```

Comment:
Not sure what our typical convention is, but it seems like over half of the code in `execute_batches_internal` now deals w/ collection/recording/processing of metrics. The collection part obviously needs to live here, but I'm wondering if the processing/recording piece could be shoved into its own function (either called from here or one level up in `execute_batches`).

Maybe create the hashmap up there and pass a reference to be filled out by the threads. I think we could even avoid passing down `confirmation_timing`

Reply:
Makes sense, I refactored the code such that `execute_batches_internal ` now returns all the metrics it collects in a struct `ExecuteBatchesInternalMetrics` which is then aggregated in `execute_batches()` via a  separate function `process_execute_batches_internal_metrics()` as you suggested: https://github.com/solana-labs/solana/pull/25752/commits/ef9608d9d799ece8405799331b078fd49e29e82b